newIDE/app/src/Utils/GDevelopServices/User.js

```diff
@@ -48,3 +75,68 @@ export const getUserPublicProfile = (
     .get(`${GDevelopUserApi.baseUrl}/user-public-profile/${id}`)
     .then(response => response.data);
 };
+
+const postBadgeIfLocked = (
+  authenticatedUser: AuthenticatedUser,
+  achievementId: string
+): ?Promise<?Badge> => {
+  const {
+    badges,
+    firebaseUser,
+    getAuthorizationHeader,
+    onBadgesChanged,
+  } = authenticatedUser;
+  if (!badges || !firebaseUser) return null;
+  if (badges.map(badge => badge.achievementId).includes(achievementId)) {
+    return null;
+  }
+  console.log(`posting achievement ${achievementId}`);
+
+  const userId = firebaseUser.uid;
+  return getAuthorizationHeader().then(authorizationHeader =>
+    axios
+      .post(
+        `${GDevelopUserApi.baseUrl}/user/${userId}/badge`,
+        {
+          achievementId,
+        },
+        {
+          params: {
+            userId,
+          },
+          headers: {
+            Authorization: authorizationHeader,
+          },
+        }
+      )
+      .then(response => {
+        onBadgesChanged();
+        return response.data;
+      })
+      .catch(err => {
+        if (err.response.status === 409) {
+          console.warn('Badge already exists');
+        } else {
+          throw err;
+        }
+      })
+  );
+};
+
+export const addPostBadgePreHookIfLocked = (
+  authenticatedUser: AuthenticatedUser,
+  achievementId: string,
+  callback: Function
+): Function => {
+  const { badges } = authenticatedUser;
+  if (!badges) return callback;
+  if (badges.map(badge => badge.achievementId).includes(achievementId)) {
```

Comment:
aren't you doing this logic twice? here and in the `postBadgeIfLocked` function

Reply:
Yes, I can put it in common: but the first is to add the hook or not (and prevent code execution if badge is already owned), the other is to make the call or not, in case the hook is still set.

Reply:
I think it makes sense, you can both:
- add a comment to tell "Check if the badge is owned now, to avoid executing any extra code if the badge is already owned".
- Put `badges.map(badge => badge.achievementId).includes(achievementId)` in a helper function `hasBadgeWithAchievementId` to make this a tiny bit less cryptic.